### PR TITLE
jewel: rbd: Potential race when removing two-way mirroring image

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/18556

### CODE REVIEW

**Comment by smithfarm** [comment]
@rdias Can you look at the build failure?

**Comment by smithfarm** [comment]
@rjfd Can you look at the build failure?

**Comment by rjfd** [comment]
@smithfarm the problem is due to some code added by #10980. That PR adds an asynchronous version of  `Journal::is_tag_owner` method, which is not available in Jewel.
I see to solutions to this problem:

a. change the code to use the synchronous version of `Journal::is_tag_owner`
b. add the `Journal::is_tag_owner` asynchronous method to Jewel

@dillaman what solution you think it's preferable? I can do either.

**Comment by dillaman** [comment]
I'm fine w/ cherry-picking 953561f84140efcc870ec4a8f1ecb732b8e99409 to jewel to resolve the build failure -- just do not want to pull in the entire PR

**Comment by smithfarm** [comment]
cherry-pick done, waiting for build

File: src/librbd/Journal.cc [hunk 1]
```diff
@@ -170,6 +170,167 @@ class SafeTimerSingleton : public SafeTimer {
   }
 };
 
+template <typename I>
+struct C_IsTagOwner : public Context {
+  I *image_ctx;
+  bool *is_tag_owner;
+  Context *on_finish;
+
+  Journaler *journaler;
```
Comment (rjfd) [inline src/librbd/Journal.cc]: Add add `using ::journal::Journaler;` just before this line

**Comment by smithfarm** [comment]
@rjfd When I merge this into the integration branch `ceph-ci/wip-jewel-backports` it fails to compile:

```
tools/rbd_mirror/image_replayer/BootstrapRequest.cc:7:30: fatal error: IsPrimaryRequest.h: No such file or directory
compilation terminated.
Makefile:22988: recipe for target 'tools/rbd_mirror/image_replayer/BootstrapRequest.lo' failed
```

Full build log at https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=xenial,DIST=xenial,MACHINE_SIZE=huge/1229//consoleFull

**Comment by smithfarm** [comment]
@rjfd Nevermind, I neglected to add `tools/rbd_mirror/image_replayer/IsPrimaryRequest.h` to `noinst_HEADERS` in `Makefile-client.am` - will send to Shaman again.

**Comment by dachary** [comment]
@dillaman does this backport look good to merge ? It passed the rbd suite http://tracker.ceph.com/issues/17851#note-67 except for two intermittent failures that are, I believe, unrelated.

**Pull request merged**